sql/core/src/test/scala/org/apache/spark/sql/streaming/ContinuousQueryManagerSuite.scala

```diff
@@ -0,0 +1,183 @@
+
+      val q1 = stopRandomQueryAsync(500 milliseconds)
+      testAwaitAnyTermination(ExpectNotBlocked, expectedReturnedValue = q1)
+      eventually(Timeout(streamingTimout)) { require(!q1.isActive) }
+
+
+      val q2 = stopRandomQueryAsync(500 milliseconds)
+      testAwaitAnyTermination(
+        ExpectNotBlocked,
+        awaitTimeout = 2 seconds,
+        expectedReturnedValue = Some(q2),
+        testTimeout = 4 seconds)
+      eventually(Timeout(streamingTimout)) { require(!q2.isActive) }
+
+      val q3 = stopRandomQueryAsync(500 milliseconds)
+      testAwaitAnyTermination(
+        ExpectNotBlocked, awaitTimeout = 100 milliseconds, expectedReturnedValue = None)
+      eventually(Timeout(streamingTimout)) { require(!q3.isActive) }
+
+      assert(sqlContext.streams.active.size === 2)
+    }
+  }
+
+  private def withQueriesOn(datasets: Dataset[_]*)(body: Seq[ContinuousQuery] => Unit): Unit = {
+    failAfter(streamingTimout) {
+      val queries = withClue("Error starting queries") {
+        datasets.map { ds =>
+          @volatile var query: StreamExecution = null
+          try {
+            val df = ds.toDF
+            query = sqlContext
+              .streams
+              .startQuery(StreamExecution.nextName, df, new MemorySink(df.schema))
+              .asInstanceOf[StreamExecution]
+          } catch {
+            case NonFatal(e) =>
+              if (query != null) query.stop()
+              throw e
+          }
+          query
+        }
+      }
+      try {
+        body(queries)
+      } finally {
+        queries.foreach(_.stop())
+      }
+    }
+  }
+
+  private def makeDataset: (MemoryStream[Int], Dataset[Int]) = {
+    val inputData = MemoryStream[Int]
+    val mapped = inputData.toDS.map(6 / _)
+    (inputData, mapped)
+  }
+
+  private def testAwaitAnyTermination(
+      expectedBehavior: ExpectedBehavior,
+      expectedReturnedValue: Any = null,
+      awaitTimeout: Span = null,
+      testTimeout: Span = 1000 milliseconds
+    ): Unit = {
+    require(expectedBehavior != ExpectNotBlocked || expectedReturnedValue != null,
+      "Expected returned value not specified when awaitTermination is expected to be not blocked")
+
+    def awaitTermFunc(): Unit = {
+      val returnedValue = if (awaitTimeout != null && awaitTimeout.toMillis > 0) {
+        sqlContext.streams.awaitAnyTermination(awaitTimeout.toMillis)
+      } else {
+        sqlContext.streams.awaitAnyTermination()
+      }
+      assert(returnedValue === expectedReturnedValue,
+        "Returned value does not match expect3ed")
+    }
+
+    AwaitTerminationTester.test(expectedBehavior, awaitTermFunc, testTimeout)
+  }
+
+  private def stopRandomQueryAsync(delay: Span): ContinuousQuery = {
+    import scala.concurrent.ExecutionContext.Implicits.global
+
+    val activeQueries = sqlContext.streams.active
+    val queryToStop = activeQueries(Random.nextInt(activeQueries.length))
```

Comment:
Add a log about the query to stop? Otherwise, it may be hard to debug if we don't know the value of `Random.nextInt(activeQueries.length)`.